utils/yoke/src/kinda_sorta_dangling.rs

```diff
@@ -0,0 +1,83 @@
+
+/// This type is intended to be similar to the type `MaybeDangling<T>`
+/// proposed in [RFC 3336].
+///
+/// The effect of this is that in Rust's safety model, types inside here are not
+/// expected to have any memory dependent validity properties (`dereferenceable`, `noalias`).
+///
+/// See [#3696] for a testcase where `Yoke` fails this under miri's field-retagging mode.
+///
+/// This has `T: 'static` since we don't need anything
+/// else and we don't want to have to think (more) about variance.
+///
+/// After [RFC 3336] lands we can use `MaybeDangling` instead.
+///
+/// [RFC 3336]: https://github.com/rust-lang/rfcs/pull/3336
+/// [#3696]: https://github.com/unicode-org/icu4x/issues/3696
+#[repr(transparent)]
+pub(crate) struct KindaSortaDangling<T: 'static> {
+    /// Safety invariant: This is always an initialized T, never uninit or other
+    /// invalid bit patterns. Its drop glue will execute during Drop::drop rather than
+    /// during the drop glue for KindaSortaDangling, which is
+    dangle: MaybeUninit<T>,
+}
+
+impl<T: 'static> KindaSortaDangling<T> {
+    #[inline]
+    pub(crate) const fn new(dangle: T) -> Self {
+        KindaSortaDangling {
+            dangle: MaybeUninit::new(dangle),
+        }
+    }
+    #[inline]
+    pub(crate) fn into_inner(self) -> T {
+        // Self has a destructor, we want to avoid having it be called
+        let manual = ManuallyDrop::new(self);
+        // Safety:
+        // We can call assume_init_read() due to the library invariant on this type,
+        // however since it is a read() we must be careful about data duplication.
+        // The only code using `self` after this is the drop glue, which we have disabled via
+        // the ManuallyDrop.
+        unsafe { manual.dangle.assume_init_read() }
+    }
+}
+
+impl<T: 'static> Deref for KindaSortaDangling<T> {
+    type Target = T;
+    #[inline]
+    fn deref(&self) -> &T {
+        // Safety: Safe due to the safety invariant on `dangle`;
+        // we can always assume initialized
+        unsafe { self.dangle.assume_init_ref() }
+    }
+}
+
+impl<T: 'static> DerefMut for KindaSortaDangling<T> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut T {
+        // Safety: Safe due to the safety invariant on `dangle`;
+        // we can always assume initialized
+        unsafe { self.dangle.assume_init_mut() }
+    }
+}
+
+impl<T: 'static> Drop for KindaSortaDangling<T> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            // Safety: We are reading and dropping a valid initialized T.
+            //
+            // As read() is a duplication operation we must be careful that the original value isn't
+            // used afterwards. It won't be because this is drop and the only
+            // code that will run after this is `self`'s drop glue, and that drop glue is empty
+            // because MaybeUninit has no drop.
+            let _drop = self.dangle.assume_init_read();
```

Comment:
Not sure how this could be problematic, types generally must be movable after all.

Reply:
I admint maybe the current "drop by decay" approach is fine, but I am quite confident that `drop(….assume_init_read())` would be unsound in face of dodgy but ultimately legitimate code:

```rs
fn main() {
    let it: &mut (i32, &i32) = &mut *MaybeDangling::new((42, &27));
    it.1 = unsafe {
        // Safety (expectation):
        //   - nobody accesses the `.1` field in this `fn` (`*it` is no longer moved).
        //   - otherwise, `MaybeDangling`'s design promises not to cause
        //     problems with the `dereferenceable`ity of `it.1` (lack thereof).
        &*(&it.0 as *const i32)
    };
    dbg!(&it); // OK
} // <- Reality: `MaybeDangling` is dropped here;
  // if it runs `drop(….read())` then it asserts `dereferenceable`ity: UB!
```

Finally, note that when the `Drop` impl of `MaybeDangling` becomes `<#[may_dangle] T>`, then this snippet does not even need `unsafe` on the caller's end:

```rs
// No `unsafe`!
it.1 = &it.0;
```

So the rule of thumb is: "types generally must be movable" is not a property that can hold in the presence of `#[may-dangle]`-drops, and if there is one type which screams `#[may_dangle]`-drop more than anyone else it is `MaybeDangling<T>`, hence why even without such an `unsafe impl Drop` we could nonetheless expect people to eventually write this kind of `unsafe` code

Reply:
Yeah I'm fine being safe vs sorry, I'll make this change.

Reply:
Hm, yeah okay that is fair. And anyway dropping in-place is also more efficient.

(But packed structs do a move-on-drop for their fields and I wonder if that could be an issue still... I guess it's fine since then the type will still be `MaybeDangling` which doesn't assert noalias/dereferenceable. It's only once we get to the underlying `T` that we have to be careful.)